### ISSUE TITLE
Support Provider-Defined Functions (Plugin Framework)

### ISSUE THREAD

**Issue by t0yv0** [pulumi/pulumi-terraform-bridge#1955]
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

See https://developer.hashicorp.com/terraform/plugin/framework/functions/concepts

**Comment by iwahbe** [comment]
## Acceptance criteria

Provider-defined functions are a Plugin Framework / protocol 5.5+/6.5+ feature (Terraform 1.8+). In Terraform they live in a namespace fully separate from resources and data sources: names are registered unprefixed (e.g. `parse_arn`, not `aws_parse_arn`) and called as `provider::<local-name>::<function-name>(...)`. They are pure, offline computations — no side effects, no access to provider configuration. The bridge maps them to Pulumi invokes.

### Build-time (tfgen)

- [ ] Functions reported by a PF provider's `GetFunctions`/`GetProviderSchema` appear in the generated Pulumi schema as entries in the `functions` map, with SDKs generated for all supported languages.
- [ ] `ProviderInfo` gains a mapping surface for functions, parallel to `DataSources`, but keyed by the unprefixed TF function name. Automatic token computation (`MustComputeTokens`) covers functions, and a token strategy guarantees no collision with data-source invoke tokens (a TF function `foo` and a data source mapped to `getFoo` must coexist).
- [ ] Missing/extra mapping validation applies to functions with the same semantics as resources and data sources (`PULUMI_SKIP_MISSING_MAPPING_ERROR`, `PULUMI_SKIP_EXTRA_MAPPING_ERROR`).
- [ ] Generated functions use the modern `FunctionSpec` shape: ordered TF positional parameters project via `multiArgumentInputs` (TF parameter `Name` is documentation-only, so duplicate/empty names must be handled deterministically), and the TF `Return` type projects via `returnType` as a direct (non-object) return. A trailing TF variadic parameter projects to a final list-typed argument, since the Pulumi schema has no variadic concept.
- [ ] All supported languages implement `multiArgumentInputs` before this ships. Today only Node.js (SDK-gen + program-gen) and .NET (SDK-gen only) support it; Go and Python SDK-gen fail with a hard error when it is set, Java SDK-gen ignores it, and Go/Python/Java program-gen error (an earlier revision of these criteria incorrectly claimed unsupported languages fall back to the single-args-object form). This issue is blocked on the per-language tracking issues: pulumi/pulumi#23548 (Go), pulumi/pulumi#23549 (Python), pulumi/pulumi-java#2225 (Java), pulumi/pulumi-dotnet#1039 (.NET program-gen).
- [ ] `Summary`, `Description`, and `DeprecationMessage` flow into the generated schema docs and deprecation metadata.

### Runtime

- [ ] Calling the generated invoke routes through the bridged provider to the TF `CallFunction` RPC, converting Pulumi property values to `tftypes` argument values (re-packing the variadic list) and the result back.
- [ ] Functions are callable without provider configuration having been applied (TF enforces config-independence; the bridge must not require `Configure` semantics that TF functions don't need).
- [ ] `FunctionError` (including argument-scoped errors) surfaces as a useful Pulumi error naming the function and, where available, the offending argument.
- [ ] Null handling honors `AllowNullValue`: missing/nil args are rejected with a clear error unless the parameter allows null. (`AllowUnknownValues` is intentionally out of play: Pulumi resolves invoke inputs before calling, so the bridge always calls with known values.)
- [ ] Muxed providers dispatch function invokes to the correct sub-provider, and function tokens participate in muxer mapping checks.
- [ ] The dynamic bridge (`dynamic/*`) exposes provider-defined functions from dynamically bridged PF providers, end to end (schema discovery through `CallFunction`).

### Conversion and docs

- [ ] `pkg/convert` translates `provider::<name>::<fn>(...)` call syntax in HCL examples to the corresponding Pulumi invoke in generated docs/examples, or degrades gracefully (clear conversion failure, not silent corruption).
- [ ] Registry docs for functions are generated from the TF function documentation.

### Testing

- [ ] Schema-generation golden tests covering: a function with positional params, a variadic function, a function whose name would collide with a data-source token, and a deprecated function.
- [ ] Cross-tests (`pkg/pf/tests/internal/cross-tests`) validating that calling the bridged invoke produces the same result Terraform produces for the same function and arguments, including error cases.
- [ ] A muxed-provider test exercising a function invoke through the muxer.
- [ ] A dynamic-bridge test exercising a function on a real upstream provider that defines them.

### Out of scope

- SDKv2-based providers: the protocol predecessor has no function support; SDKv2-only providers cannot expose functions. Muxed providers get them via their PF half only.
- Terraform 1.x built-in functions and module-defined functions — only provider-defined functions.

**Comment by Frassle** [comment]
> They are pure, offline computations — no side effects, no access to provider configuration. The bridge maps them to Pulumi invokes.

Is that true? I thought they could be stateful that's why terraform treats plan and apply separate for them.

**Comment by iwahbe** [comment]
> > They are pure, offline computations — no side effects, no access to provider configuration. The bridge maps them to Pulumi invokes.
> 
> Is that true? I thought they could be stateful that's why terraform treats plan and apply separate for them.

Yes. From the TF framework docs: https://developer.hashicorp.com/terraform/plugin/framework/functions/concepts

> The purpose of provider-defined functions is to encapsulate offline, computational logic beyond Terraform's built-in functions to simplify practitioner configurations. Terraform expects that provider-defined functions are implemented without side-effects and as pure functions where given the same input data that they always return the same output.

It's possible you are thinking of [actions](https://developer.hashicorp.com/terraform/plugin/framework/actions), which are definitionally a way for a provider to expose non-resource side-effects.

**Comment by Frassle** [comment]
Ahhh yeh I got those muddled up, I think the branch I pinged you was a start on 'actions' as well.
